File: backends/node/shared/implementations/caching/distributed/redis/src/handlers/r/get-ttl.ts
import type Redis from "ioredis";
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result } from "@d2/result";
import { redisErrorResult } from "../../redis-error-result.js";
import type { DistributedCache } from "@d2/interfaces";

type Input = DistributedCache.GetTtlInput;
type Output = DistributedCache.GetTtlOutput;

export class GetTtl extends BaseHandler<Input, Output> implements DistributedCache.IGetTtlHandler {
  private readonly redis: Redis;

  constructor(redis: Redis, context: IHandlerContext) {
    super(context);
    this.redis = redis;
  }

  protected async executeAsync(input: Input): Promise<D2Result<Output | undefined>> {
    try {
      const pttl = await this.redis.pttl(input.key);

      // -2 = key doesn't exist, -1 = no expiry
      const timeToLiveMs = pttl > 0 ? pttl : undefined;
      return D2Result.ok({ data: { timeToLiveMs } });
    } catch (err: unknown) {
      return redisErrorResult(err);
    }
  }
}

File: backends/dotnet/shared/Implementations/Caching/Distributed/DistributedCache.Redis/Handlers/R/GetTtl.cs
// -----------------------------------------------------------------------
// <copyright file="GetTtl.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.DistributedCache.Redis.Handlers.R;

using D2.Shared.Handler;
using D2.Shared.Result;
using Microsoft.Extensions.Logging;
using StackExchange.Redis;
using S = D2.Shared.Interfaces.Caching.Distributed.Handlers.R.IRead;

/// <summary>
/// Handler for getting the time-to-live of a key in the distributed cache.
/// </summary>
public partial class GetTtl : BaseHandler<S.IGetTtlHandler, S.GetTtlInput, S.GetTtlOutput>,
    S.IGetTtlHandler
{
    private readonly IConnectionMultiplexer r_redis;

    /// <summary>
    /// Initializes a new instance of the <see cref="GetTtl"/> class.
    /// </summary>
    ///
    /// <param name="redis">
    /// The Redis connection multiplexer.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public GetTtl(
        IConnectionMultiplexer redis,
        IHandlerContext context)
        : base(context)
    {
        r_redis = redis;
    }

    /// <inheritdoc/>
    protected override async ValueTask<D2Result<S.GetTtlOutput?>> ExecuteAsync(
        S.GetTtlInput input,
        CancellationToken ct = default)
    {
        try
        {
            var db = r_redis.GetDatabase();
            var ttl = await db.KeyTimeToLiveAsync(input.Key);

            return D2Result<S.GetTtlOutput?>.Ok(
                new S.GetTtlOutput(ttl));
        }
        catch (RedisException ex)
        {
            LogGetTtlFailed(Context.Logger, ex, input.Key, TraceId);

            return D2Result<S.GetTtlOutput?>.ServiceUnavailable(traceId: TraceId);
        }

        // Let the base handler catch any other exceptions.
    }

    /// <summary>
    /// Logs that a Redis exception occurred while getting the TTL of a key.
    /// </summary>
    [LoggerMessage(EventId = 1, Level = LogLevel.Error, Message = "RedisException occurred while getting TTL for key '{Key}'. TraceId: {TraceId}")]
    private static partial void LogGetTtlFailed(ILogger logger, Exception ex, string key, string? traceId);
}
